### DISCUSSION TITLE
Racy access to ListDepth in CxPlatPoolFree in quic_platform_posix.h

### DISCUSSION THREAD

**Discussion by GaneshRapolu** [microsoft/msquic#2766] [General]
See https://github.com/microsoft/msquic/blob/1f2e91a5fa1459cde259768b746bb70743b5f968/src/inc/quic_platform_posix.h#L537

Pool->ListDepth is accessed with and without protection of Pool->Lock. However, ListDepth is declared as uint16_t in struct CXPLAT_POOL

https://github.com/microsoft/msquic/blob/1f2e91a5fa1459cde259768b746bb70743b5f968/src/inc/quic_platform_posix.h#L417

It is neither an atomic variable, nor is the access inside CxPlatPoolFree atomic. Thus the read of ListDepth outside of Pool->Lock races with the write of ListDepth inside of Pool->Lock in CxPlatPoolFree. From my understanding, this is UB.

Think we can fix this without perf impact using relaxed atomics, as the CXPLAT_POOL_MAXIMUM_DEPTH is a best effort bound anyway.

**Comment by nibanks** [comment]
This depth really is treated more as a guideline than a hard limit. This is just used to try to impose a reasonable limit on the maximum number of items in the pool, but the limit is still arbitrary; so there's not a lot of reason to add complexity to enforce it. Unless there is actually an issue that can be reproduced related to this, I'd most likely leave it as is.

Thanks for doing the work to analyze the code and provide this feedback!

**Reply by GaneshRapolu** [reply]
Thanks for the quick response. I am aware that it is not and is not intended to be a hard limit. The change that I am proposing would also not make it a hard limit.

My only concern is that the racy access to ListDepth is undefined behavior (UB) and that (in theory) can cause arbitrary miscompilation/runtime behavior. To avoid this UB, we can use relaxed atomics to access ListDepth. Using relaxed atomics should not add any perf impact (ex. just mov on x86). Relaxed atomics will also not turn this into a hard limit.

So using functions from https://en.cppreference.com/w/c/atomic , the code would be something like

declare ListDepth as

```
atomic_uint_fast16_t ListDepth; 
```

change the portion of CxPlatPoolFree to be
```
    if (atomic_load_explicit(&Pool->ListDepth, memory_order_relaxed) >= CXPLAT_POOL_MAXIMUM_DEPTH) {
        CxPlatFree(Entry, Pool->Tag);
    } else {
        CxPlatLockAcquire(&Pool->Lock);
        CxPlatListPushEntry(&Pool->ListHead, (CXPLAT_SLIST_ENTRY*)Entry);
        uint16_t listDepth = atomic_load_explicit(&Pool->ListDepth, memory_order_relaxed);
        atomic_store_explicit(&Pool->ListDepth,  listDepth + 1,  memory_order_relaxed);
        CxPlatLockRelease(&Pool->Lock);
    }
```

Ditto for the other uses of ListDepth, ex. the write to ListDepth in CxPlatPoolAlloc

https://github.com/microsoft/msquic/blob/1f2e91a5fa1459cde259768b746bb70743b5f968/src/inc/quic_platform_posix.h#L508

**Reply by nibanks** [reply]
Feel free to make a PR to try to improve the code, but we've never had a problem, and if the fix results in any significant performance regressions, I'd be hesitant to take it.
